File: packages/protocol-schema/src/rpc/envelope.ts
import { Type, type Static } from "@sinclair/typebox";

export const RpcIdSchema = Type.Union([Type.String(), Type.Integer()], {
  $id: "RpcId",
});
export type RpcId = Static<typeof RpcIdSchema>;

export const RpcErrorDataSchema = Type.Object(
  {
    code: Type.Integer(),
    message: Type.String(),
    data: Type.Optional(Type.Unknown()),
  },
  { $id: "RpcErrorData" }
);
export type RpcErrorData = Static<typeof RpcErrorDataSchema>;

export const RpcRequestEnvelopeSchema = Type.Object(
  {
    id: RpcIdSchema,
    method: Type.String(),
    params: Type.Optional(Type.Unknown()),
  },
  { $id: "RpcRequestEnvelope" }
);
export type RpcRequestEnvelope = Static<typeof RpcRequestEnvelopeSchema>;

export const RpcResponseEnvelopeSchema = Type.Object(
  {
    id: RpcIdSchema,
    result: Type.Optional(Type.Unknown()),
    error: Type.Optional(RpcErrorDataSchema),
  },
  { $id: "RpcResponseEnvelope" }
);
export type RpcResponseEnvelope = Static<typeof RpcResponseEnvelopeSchema>;

export const RpcNotificationEnvelopeSchema = Type.Object(
  {
    method: Type.String(),
    params: Type.Optional(Type.Unknown()),
  },
  { $id: "RpcNotificationEnvelope" }
);
export type RpcNotificationEnvelope = Static<
  typeof RpcNotificationEnvelopeSchema
>;

export const RpcMessageEnvelopeSchema = Type.Union(
  [
    RpcRequestEnvelopeSchema,
    RpcResponseEnvelopeSchema,
    RpcNotificationEnvelopeSchema,
  ],
  { $id: "RpcMessageEnvelope" }
);
export type RpcMessageEnvelope = Static<typeof RpcMessageEnvelopeSchema>;

File: crates/canywhere-protocol/src/rpc/envelope.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(untagged)]
pub enum RpcId {
    Number(i64),
    String(String),
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct RpcErrorData {
    pub code: i32,
    pub message: String,
    pub data: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct RpcRequestEnvelope {
    pub id: RpcId,
    pub method: String,
    pub params: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct RpcResponseEnvelope {
    pub id: RpcId,
    pub result: Option<serde_json::Value>,
    pub error: Option<RpcErrorData>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct RpcNotificationEnvelope {
    pub method: String,
    pub params: serde_json::Value,
}
